File: 9.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type pos struct{ x, y int }

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func direction(from, to int) int {
	if to == from {
		return 0
	} else if to > from {
		return 1
	}
	return -1
}

type rope struct {
	knots   []pos
	visited map[pos]bool
}

func (r rope) move(dir string) {
	switch dir {
	case "R":
		r.knots[0].x += 1
	case "L":
		r.knots[0].x -= 1
	case "D":
		r.knots[0].y += 1
	case "U":
		r.knots[0].y -= 1
	}
	for i := 1; i < len(r.knots); i++ {
		if abs(r.knots[i-1].x-r.knots[i].x) > 1 || abs(r.knots[i-1].y-r.knots[i].y) > 1 {
			r.knots[i] = pos{
				r.knots[i].x + direction(r.knots[i].x, r.knots[i-1].x),
				r.knots[i].y + direction(r.knots[i].y, r.knots[i-1].y),
			}
		}
	}
	r.visited[r.knots[len(r.knots)-1]] = true
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	rope1 := rope{[]pos{{0, 0}, {0, 0}}, map[pos]bool{}}
	rope2 := rope{[]pos{{0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}}, map[pos]bool{}}
	for scanner.Scan() {
		var dir string
		var n int
		fmt.Sscanf(scanner.Text(), "%s %d", &dir, &n)
		for i := 0; i < n; i++ {
			rope1.move(dir)
			rope2.move(dir)
		}
	}
	fmt.Println(len(rope1.visited), len(rope2.visited))
}

File: 9.swift
struct Pos: Hashable {
    var x = 0, y = 0
}

struct Rope {
    var knots: [Pos] = []
    var visited: Set<Pos> = []

    init(_ nknots: Int) {
        for _ in 1...nknots { self.knots.append(Pos()) }
    }

    mutating func move(_ dx: Int, _ dy: Int) {
        knots[0].x += dx
        knots[0].y += dy
        for i in 1..<knots.count {
            if abs(knots[i - 1].x - knots[i].x) > 1 ||
               abs(knots[i - 1].y - knots[i].y) > 1 {
                knots[i].x += direction(knots[i].x, knots[i - 1].x)
                knots[i].y += direction(knots[i].y, knots[i - 1].y)
            }
        }
        visited.insert(knots.last!)
    }

    func direction(_ from: Int, _ to: Int) -> Int {
        return from == to ? 0 : (from > to ? -1 : 1)
    }
}

var rope1 = Rope(2), rope2 = Rope(10)
while let line = readLine() {
    var dx = 0, dy = 0, n = Int(line.dropFirst(2))!
    switch line.prefix(1) {
        case "R": dx = 1
        case "L": dx = -1
        case "D": dy = 1
        case "U": dy = -1
        default: break
    }
    for _ in 1...n { 
        rope1.move(dx, dy) 
        rope2.move(dx, dy) 
    }
}
print(rope1.visited.count, rope2.visited.count)